src/Paginator.php

```diff
@@ -0,0 +1,175 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Rebilly;
+
+use ArrayObject;
+use Countable;
+use Iterator;
+use RuntimeException;
+use OutOfBoundsException;
+
+/**
+ * Class Paginator.
+ *
+ * @author Veaceslav Medvedev <veaceslav.medvedev@rebilly.com>
+ * @version 0.1
+ */
+final class Paginator implements Iterator, Countable
+{
+    const DEFAULT_SIZE = 100;
+
+    /** @var Client */
+    private $client;
+
+    /** @var string */
+    private $uri;
+
+    /** @var array */
+    private $params = [];
+
+    /** @var int */
+    private $limit = self::DEFAULT_SIZE;
+
+    /** @var int */
+    private $offset = 0;
+
+    /** @var int */
+    private $totalItems;
+
+    /** @var Rest\Collection */
+    private $currentSegment;
+
+    /**
+     * @param Client $client
+     * @param string $uri
+     * @param array|ArrayObject $params
+     */
+    public function __construct(Client $client, $uri, $params = [])
+    {
+        $this->client = $client;
+        $this->uri = $uri;
+        $this->params = (array) $params + ['limit' => $this->limit, 'offset' => $this->offset];
+
+        $this->load($this->params);
+    }
+
+    /**
+     * @param array $params
+     */
+    private function load(array $params)
+    {
+        /** @var Rest\Collection $segment */
+        $segment = $this->client->get($this->uri, $params);
+
+        if ($segment->getOffset() % $segment->getLimit() !== 0) {
+            throw new RuntimeException('Not relevant segment offset and limit');
+        }
+
+        $this->currentSegment = $segment;
+        $this->totalItems = $segment->getTotalItems();
+        $this->offset = $segment->getOffset();
+        $this->limit = $segment->getLimit();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @return Rest\Collection
+     */
+    public function current()
+    {
+        if ($this->currentSegment === null) {
+            if (!$this->valid()) {
+                throw new OutOfBoundsException('Cannot load segment, invalid offset');
+            }
+
+            $this->load(['offset' => $this->offset] + $this->params);
+        }
+
+        return $this->currentSegment;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function next()
+    {
+        $this->currentSegment = null;
+        $this->offset += $this->limit;
+    }
+
+    /**
+     * Move backward to previous page.
+     */
+    public function previous()
+    {
+        $this->currentSegment = null;
+        $this->offset -= $this->limit;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function key()
+    {
+        return (int) ceil($this->offset / $this->limit) + 1;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function valid()
+    {
+        return $this->offset >= 0 && $this->offset <= $this->totalItems;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function rewind()
+    {
+        $this->currentSegment = null;
+        $this->totalItems = null;
```

Comment:
rewind() is to get to the starting point... as I explained above with totalItems set to null methods that use it would give an error...

totalItes is set only in `load()` but then this would not be a rewind(), it would be a reload() or refresh()